src/test/java/net/satago/tapestry5/jpa/test/TopLevelServiceImpl.java

```diff
@@ -0,0 +1,95 @@
+/**
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package net.satago.tapestry5.jpa.test;
+
+import javax.persistence.EntityManager;
+
+import net.satago.tapestry5.jpa.EntityTransactionManager;
+import net.satago.tapestry5.jpa.test.entities.ThingOne;
+import net.satago.tapestry5.jpa.test.entities.ThingTwo;
+
+import org.apache.tapestry5.ioc.Invokable;
+import org.apache.tapestry5.jpa.annotations.CommitAfter;
+
+public class TopLevelServiceImpl implements TopLevelService {
+
+	private final EntityManager em;
+	private final NestedService nestedService;
+    private final EntityTransactionManager entityTransactionManager;
+
+    public TopLevelServiceImpl(EntityManager em, NestedService nestedService,
+            EntityTransactionManager transactionalUnits)
+    {
+		this.em = em;
+		this.nestedService = nestedService;
+		this.entityTransactionManager = transactionalUnits;
+	}
+
+	@Override
+	@CommitAfter
+	public void createThingOneAndTwo(String nameOne, String nameTwo) {
+		ThingOne thingOne = new ThingOne();
+		thingOne.setName(nameOne);
+		em.persist(thingOne);
+        nestedService.createThingTwo(nameTwo);
+	}
+
+    @Override
+    @CommitAfter
+    public void createThingOneThenTwo(final String nameOne, final String nameTwo)
+    {
+        entityTransactionManager.invokeAfterCommit(null, new Invokable<Boolean>()
+        {
+            @Override
+            public Boolean invoke()
+            {
+                // nestedService.createThingTwo(nameTwo);
+                ThingTwo thingTwo = new ThingTwo();
+                thingTwo.setName(nameTwo);
+                em.persist(thingTwo);
```

Comment:
This test is not accurate, it's not clear what should happen here because there's no active transaction.

The `afterCommit` callback executed... after commit, outside of transaction scope, no new transaction will be open. And in JPA there should always be an open transaction, otherwise the behavior is not defined.

http://stackoverflow.com/a/21673304